onnxruntime/contrib_ops/webgpu/quantization/dp4a_matmul_nbits.cc

```diff
@@ -258,11 +258,120 @@ Status DP4AMatMulNBitsProgram::GenerateShaderCode(ShaderHelper& shader) const {
+  shader.AddInput("scales_a", ShaderUsage::UseUniform);
+  shader.AddInput("input_b", ShaderUsage::UseUniform);
+  shader.AddInput("scales_b", ShaderUsage::UseUniform);
+  shader.AddOutput("output", ShaderUsage::UseUniform | ShaderUsage::UseElementTypeAlias);
+
+  shader.AdditionalImplementation() << R"ADDNL_FN(
+    const tile_size = 16u; // tile_size = tile_size_vec * output components
+    const tile_size_vec = 4u;
+    const tile_size_k_vec = 16u; // tile_size_vec * tile_size_k_vec = workgroup size
+    // Shared memory
+    var<workgroup> tile_A : array<vec4<u32>, 32>;                    // 512 scalars
+    var<workgroup> scale_A : array<output_element_t, 4>;             // 4
+    var<workgroup> inter_results: array<array<vec4<output_element_t>, tile_size_k_vec>, tile_size_vec>;
+    fn loadSHMA(a_global:u32, kidx_v:u32, col: u32)
+    {
+      let k_offset = kidx_v + col;
+      if (k_offset >= uniforms.K16) {
+        return;
+      }
+
+      tile_A[col] = input_a[a_global*uniforms.K16+k_offset];
+      if (col < 4)
+      {
+        // kidx_v - covers 16 values of k
+        scale_A[col] = scales_a[a_global*(uniforms.K/128) + k_offset/8 + col];
+      }
+    }
+    // Scaled dot product of 8 packed unsigned integers.
+    fn SDP8AI(a1:vec4<u32>, b1:vec4<u32>, a2:vec4<u32>, b2:vec4<u32>, scale:output_element_t) -> output_element_t
+    {
+        var local_sum = dot4I8Packed(a1[0], b1[0]);
+        local_sum += dot4I8Packed(a1[1], b1[1]);
+        local_sum += dot4I8Packed(a1[2], b1[2]);
+        local_sum += dot4I8Packed(a1[3], b1[3]);
+        local_sum += dot4I8Packed(a2[0], b2[0]);
+        local_sum += dot4I8Packed(a2[1], b2[1]);
+        local_sum += dot4I8Packed(a2[2], b2[2]);
+        local_sum += dot4I8Packed(a2[3], b2[3]);
+        return output_element_t(local_sum) * scale;
+    }
+  )ADDNL_FN";
+
+  shader.MainFunctionBody() << R"MAIN_FN(
+    let a_global = workgroup_id.y;
+    let b_global_base = workgroup_id.x * tile_size;
+    let idx = local_idx % tile_size_k_vec;
+    let idy = local_idx / tile_size_k_vec;
+    for (var kidx_v:u32 = 0; kidx_v < uniforms.K32; kidx_v+=16)
+    {
+      // Load Phase: Populate shared memory for the workgroup.
+      if (local_idx < 32)
+      {
+        loadSHMA(a_global, kidx_v * 2, local_idx);
+      }
+      workgroupBarrier();
+      var own_a: vec4<u32> = tile_A[idx*2];
+      var own_a1: vec4<u32> = tile_A[idx*2 + 1];
+      var own_scale_a: output_element_t = scale_A[idx / 4];
+      var own_b = vec4<u32>(0);
+      var own_b1 = vec4<u32>(0);
+      let k_offset = kidx_v+idx;
+      for (var i = 0u; i < 4u; i++) {
+        let b_global = b_global_base + idy * 4 + i;
+        if (b_global < uniforms.N && k_offset < uniforms.K32)
+        {
+          let b_offset = b_global*uniforms.K32+k_offset;
+          let b_value = input_b[b_offset];
+          var b_value_lower = vec4<i32>(unpack4xU8(b_value[0] & 0x0F0F0F0Fu)) - vec4<i32>(8);
```

Comment:
nit: is the dequantization of b easy to pull out into a function SDP8AI and that function could be shared between the generation and prefill shaders.

Reply:
Good suggestion. Done.